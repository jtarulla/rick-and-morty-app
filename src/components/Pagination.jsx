import React from 'react';

const Pagination = ({handlePage, info, loading}) => {
    const pageNumberIndex = info.next !== null ? info.next.search('page=') : 0;
    const justNumbers = /[^0-9+]/g;

    const getPageNumber = () => (
        info.next !== null
        ?   (
                info.next.slice(pageNumberIndex, pageNumberIndex+7)
                    .replace(justNumbers, '') - 1
            )
        : info.pages
    );

    return (
        <>
            {
                !loading &&
                    <div className="pagination">
                        <div className="pagination--counter">
                            Page {getPageNumber()} of {info.pages}<br />
                        </div> 
                        <div className="container">
                            <button
                                type="button"
                                onClick={() => handlePage(info.prev)}
                                disabled={!info.prev}
                            >
                                Previous
                            </button>
                            <button
                                type="button"
                                id="next"
                                onClick={() => handlePage(info.next)}
                                disabled={!info.next}
                            >
                                Next
                            </button>
                        </div>
                    </div>
            }
        </>
    );
}

export default Pagination;

