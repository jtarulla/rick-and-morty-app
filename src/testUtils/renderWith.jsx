import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

const MemoryRouterWithInitialRoutes = ({ children, initialRoutes }) => {
    return (
        <MemoryRouter initialEntries={initialRoutes}>
            {children}
        </MemoryRouter>
    );
};

export const renderWithRouter = (ui, options) => {
    const initialRoutes =
        options && options.initialRoutes ? options.initialRoutes : ["/"];

    return render(ui, {
            wrapper: (args) =>
                MemoryRouterWithInitialRoutes({
                    ...args,
                    initialRoutes,
                }),
                ...options,
        }
    );
}

export * from "@testing-library/react";
