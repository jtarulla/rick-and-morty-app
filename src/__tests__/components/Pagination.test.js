import '@testing-library/jest-dom/extend-expect';
import { fireEvent } from '@testing-library/react';

import { renderWithRouter } from 'testUtils/renderWith';
import Pagination from 'components/Pagination';
import HomePage from 'pages/homePage';
import { loading, info } from 'testUtils/mocks/characters'

const setup = async (isLoading = loading.false, handlePage = jest.fn) => {
    const wrapper = await renderWithRouter(
        <Pagination
            handlePage={handlePage}
            info={info}
            loading={isLoading}
        />
    );

    return {wrapper};
}

describe('Pagination component', () => {
    it('renders', async () => {
        const { wrapper } = await setup();
        const component = await wrapper.findByTestId('pagination-component');
        
        expect(component).toBeTruthy();
        expect(wrapper).toMatchSnapshot();
    });

    it('doesn\'t renders if loading is true', async () => {
        const { wrapper } = await setup(loading.true);

        expect(wrapper.length).toBeFalsy();
    });

    it('changes the page number when click on next button', async () => {
        const screen = renderWithRouter(<HomePage />);
        await setup(loading.false);
        const paginationButton = screen.getByText('next', { exact: false });

        const event = fireEvent.click(paginationButton);
        expect(event).toBe(true);
      });
})
