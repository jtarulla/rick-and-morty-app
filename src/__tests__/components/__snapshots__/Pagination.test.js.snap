// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pagination"
        data-testid="pagination-component"
      >
        <div
          class="pagination--counter"
          data-testid="pagination-counter"
        >
          Page 
          4
           of 
          42
          <br />
        </div>
        <div
          class="container"
        >
          <button
            type="button"
          >
            Previous
          </button>
          <button
            id="next"
            type="button"
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pagination"
      data-testid="pagination-component"
    >
      <div
        class="pagination--counter"
        data-testid="pagination-counter"
      >
        Page 
        4
         of 
        42
        <br />
      </div>
      <div
        class="container"
      >
        <button
          type="button"
        >
          Previous
        </button>
        <button
          id="next"
          type="button"
        >
          Next
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
